diff --git a/VVGL/Makefile b/VVGL/Makefile
index 891a305..64b7758 100644
--- a/VVGL/Makefile
+++ b/VVGL/Makefile
@@ -36,8 +36,10 @@ else ifeq ($(PLATFORM),Darwin)
 	CXX = clang
 	CPPFLAGS := -Wall -g -arch $(ARCH) -O3 -std=c++11 -stdlib=libc++ -fPIC
 	CPPFLAGS += -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk
-	CPPFLAGS += -I./include/ -DVVGL_SDK_MAC
+	CPPFLAGS += -I./include/ -DVVGL_SDK_GLFW
     CPPFLAGS += -DGL_SILENCE_DEPRECATION
+	# Add GLFW and GLEW include paths
+	CPPFLAGS += -I/opt/homebrew/include
 	OBJCPPFLAGS := $(CPPFLAGS)
 	
 	CPPFLAGS += -x c++
@@ -45,6 +47,8 @@ else ifeq ($(PLATFORM),Darwin)
 	
 	LDFLAGS := -arch $(ARCH) -lstdc++ -shared -fPIC -lobjc
 	LDFLAGS += -framework Foundation -framework ImageIO -framework OpenGL -framework IOSurface -framework CoreGraphics -framework CoreVideo
+	# Add GLFW and GLEW libraries
+	LDFLAGS += -L/opt/homebrew/lib -lglfw -lGLEW
 endif
 
 
diff --git a/VVISF/Makefile b/VVISF/Makefile
index 874a6a4..3d57904 100644
--- a/VVISF/Makefile
+++ b/VVISF/Makefile
@@ -35,8 +35,10 @@ else ifeq ($(PLATFORM),Darwin)
 	CXX = clang
 	CPPFLAGS := -Wall -g -arch $(ARCH) -O3 -std=c++11 -stdlib=libc++ -fPIC 
 	CPPFLAGS += -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk
-	CPPFLAGS += -I./include/ -DVVGL_SDK_MAC
+	CPPFLAGS += -I./include/ -DVVGL_SDK_GLFW
     CPPFLAGS += -DGL_SILENCE_DEPRECATION
+	# Add GLFW and GLEW include paths
+	CPPFLAGS += -I/opt/homebrew/include
 	OBJCPPFLAGS := $(CPPFLAGS)
 	
 	CPPFLAGS += -x c++
@@ -44,6 +46,8 @@ else ifeq ($(PLATFORM),Darwin)
 	
 	LDFLAGS := -arch $(ARCH) -lstdc++ -shared -fPIC -lobjc
 	LDFLAGS += -framework Foundation -framework ImageIO -framework OpenGL -framework IOSurface -framework CoreGraphics -framework CoreVideo
+	# Add GLFW and GLEW libraries
+	LDFLAGS += -L/opt/homebrew/lib -lglfw -lGLEW
 endif
 
 
